### PR TITLE
Improve BinaryValue performance 

### PR DESCRIPTION
**Pull request by garmin-mjames**
Improving performance related to `BinaryValue` objects.

Relies on #1502 providing a performance regression test. The `matrix_multiplier` test makes heavy use of the `BinaryValue` class.

CI results
----------
Python | master (4eaf429c67739886fab45afd647b4eeb37436694) | PR (d43dcdf359a8ce41d50d51c46fddae470b4baa89) | improvement
------- | --------------- | ----- | --------------
3.8 | [825 NS/S](https://github.com/cocotb/cocotb/runs/1564866751?check_suite_focus=true#step:22:5077) | [1982 NS/S]() | +140%

Outdated results
-----------------

Running `test_matrix_multiplier` locally:

sim | 5b71032 average | 71ff728 average | improvement
--- | ---------- | -------- | ------------- 
icarus | 1583 NS/S | 2413 NS/S | +52% !
aldec (vhdl) | 1204 NS/S | 2046 NS/S | +70% !!

These numbers are outsized due to the `BinaryValue`-focused test, but I think other tests will see at least a small performance improvement.

@andresdemski could you run your test against this?

### CODE REVIEW

File: cocotb/handle.py
Comment (garmin-mjames) [inline cocotb/handle.py]: This assumes that we can always trust the string the simulator returns from `get_signal_val_binstr()` only contains permitted characters.
The speed-up was quite significant, so I hope that's a valid assumption.

**Comment by cmarqu** [comment]
Could you add a short section for ``matrix_multiplier`` to https://github.com/cocotb/cocotb/blob/master/documentation/source/examples.rst?
I made an effort to at least mention all examples in the docs a while ago.

File: cocotb/binary.py
Comment (eric-wieser) [inline cocotb/binary.py]: What's the motivation for moving these to global public constants?

File: cocotb/handle.py
Comment (eric-wieser) [inline cocotb/handle.py]: Your metaclass above feels like crazy overkill - One simple option would be:
```python
@property
def value(self):
    return self._getvalue()
@value.setter
def value(self, v):
    self._setcachedvalue(v)
```

If you're trying to optimize this further, then there are still tidier ways to do it - I'll put up an alternate patch for this path

File: cocotb/binary.py
Comment (eric-wieser) [inline cocotb/binary.py]: This change means that users can no longer change `.binaryRepresentation` after the object is created. I don't know whether that was ever supported.

File: examples/matrix_multiplier/tests/test_matrix_multiplier.py
Comment (eric-wieser) [inline examples/matrix_multiplier/tests/test_matrix_multiplier.py]: Can you extract a helper here:
```python
def gen_random_matrix(rows, cols):
    return [[getrandbits(DATA_WIDTH) for col in range(cols)] for row in range(rows)]
```

---

It's then _maybe_ tempting to use it as
```python
factory.add_option('a_data', [
    lambda: (gen_random_matrix(A_ROWS, A_COLUMNS_B_ROWS) for i in range(NUM_SAMPLES))
])
factory.add_option('b_data', [
    lambda: (gen_random_matrix(A_COLUMNS_B_ROWS, B_COLUMNS) for i in range(NUM_SAMPLES))
])
```

File: cocotb/binary.py
Comment (garmin-mjames) [reply cocotb/binary.py]: I moved them so that they were defined before both `Resolver` and `BinaryValue` classes.
`BinaryValue` calls `resolver.resolve()` so it was a chicken/egg problem.

File: examples/matrix_multiplier/tests/test_matrix_multiplier.py
Comment (eric-wieser) [inline examples/matrix_multiplier/tests/test_matrix_multiplier.py]: Thsi becomes far more readable with a helper function:
```python
def _unpack_matrix(self, signal, rows, cols):
    if self.multi_dimension:
        return [[signal[(i * cols) + j].value for j in range(cols)] for i in range(row)]
    else:
        return signal.value
```
used as:
```python
    if self.bus.o_valid.value and in_values:
        self._recv((in_values[0], in_values[1], self._unpack_matrix(self.bus.C, A_ROWS, B_COLUMNS)))
        in_values = None

    if self.bus.i_valid.value:
        # Store BinaryValue objects
        in_values = (
            self._unpack_matrix(self.bus.A, A_ROWS, A_COLUMNS_B_ROWS),
            self._unpack_matrix(self.bus.B, A_COLUMNS_B_ROWS, B_COLUMNS),
        )
```

File: cocotb/binary.py
Comment (eric-wieser) [reply cocotb/binary.py]: It's not a chicken and egg problem though - `resolver.resolve` can use `BinaryValue._resolve_to_1` without any issue.

If you think they're clearer at the top of the file, then please rename them to start with an underscore (ie, to be private), and remove the `BinaryValue` attributes completely.

File: cocotb/binary.py
Comment (eric-wieser) [inline cocotb/binary.py]: Ah, now I see what you mean about the chicken and egg problem. You could put this on the last line of the file and it would work. `_RESOLVE_TO_ERROR` is fine by me though.

File: cocotb/binary.py
Comment (eric-wieser) [inline cocotb/binary.py]: Woa, this `RANDOM` thing is nonsense - it resolves the Xs to either all ones or all zeros!

File: cocotb/binary.py
Comment (eric-wieser) [inline cocotb/binary.py]: You might want to look at using `string.maketrans` which will likely be even faster

File: examples/matrix_multiplier/tests/test_matrix_multiplier.py
Comment (garmin-mjames) [reply examples/matrix_multiplier/tests/test_matrix_multiplier.py]: Agreed, thanks.

File: cocotb/binary.py
Comment (garmin-mjames) [reply cocotb/binary.py]: I noticed that, and I think resolving each bit is better.

File: cocotb/binary.py
Comment (eric-wieser) [inline cocotb/binary.py]: Likely faster would be:
```
    def __missing__(self, key):
        try:
             return self.resolve_x()
        except AttributeError:
            return key
    def __call__(self, match):
        return self[match.group(0)]
```

Comment (garmin-mjames) [reply cocotb/binary.py]: I also tried:
```
self.resolve = partial(self._regex.sub, self)
```
And saw a very small speedup, but it was lost in the variability of the ratio from run-to-run.

Comment (eric-wieser) [reply cocotb/binary.py]: I would recommend timing this type of thing locally in ipython, rather than as part of a full test suite.

File: examples/matrix_multiplier/tests/test_matrix_multiplier.py
Comment (garmin-mjames) [reply examples/matrix_multiplier/tests/test_matrix_multiplier.py]: The only downside to the lambdas is we lose the docstrings:
```
     0.00ns INFO     cocotb.test._my_test.0x7feabcca0860       decorators.py:253  in _advance                        Starting test: "run_test_array_001"
                                                                                                                     Description: Automatically generated test

                                                                                                                       a_data: <lambda> (No docstring supplied)
                                                                                                                       b_data: <lambda> (No docstring supplied)
```

Comment (eric-wieser) [reply examples/matrix_multiplier/tests/test_matrix_multiplier.py]: Good point - them maybe use the helper function, but use it from within `def gen_a` and `def gen_b`

File: cocotb/binary.py
Comment (ktbarrett) [reply cocotb/binary.py]: It was. Changing `binaryRepresentation` after creation would result in a different result for `get_value` and all the `conversion` functions from that point forward. That being said, it was probably never a good idea, and it isn't documented.

File: cocotb/handle.py
Comment (eric-wieser) [reply cocotb/handle.py]: See gh-1287 for a less complex solution

File: cocotb/binary.py
Comment (garmin-mjames) [reply cocotb/binary.py]: After investigating using `str.translate()`, it's much faster but it uses the translation value for all matching characters, so all `x` or `Z` become the same number.
The effect is the current behavior, where `xxxx` becomes `1111` or `0000`.

I can cleanup the `re.sub()` implementation if we really want per-bit random values.

Comment (eric-wieser) [reply cocotb/binary.py]: Since this patch is about performance, we should err on the side of keeping the current behavior.

In a future pass we could do a second pass to randomize X's.

File: cocotb/handle.py
Comment (garmin-mjames) [reply cocotb/handle.py]: Removed the commit in favor of your PR.

**Comment by garmin-mjames** [comment]
@cmarqu Added documentation of example.
@ktbarrett Fixed a few of the performance issues you pointed out.

Generally, the BinaryValue class has some serious problems with its internal representation and also big vs little endian. See issue #608 and PR #1020.

I stuck with performance improvements while trying to match current behavior.

File: cocotb/binary.py
Comment (garmin-mjames) [inline cocotb/binary.py]: I think this is only Python 3 compatible as-is (assumes Unicode).

Comment (ktbarrett) [reply cocotb/binary.py]: With #1289 already filed, it seems Eric is eager to rid us of Python 2. So I wouldn't worry about it.

File: cocotb/binary.py
Comment (eric-wieser) [inline cocotb/binary.py]: No need for `ord` I think

File: cocotb/binary.py
Comment (eric-wieser) [inline cocotb/binary.py]: This still won't actually do what you're expecting it to, `str.translate` caches the calls to `__missing__`.

File: cocotb/binary.py
Comment (garmin-mjames) [reply cocotb/binary.py]: `str.translate()` indexes into table with an ordinal, but you can return ordinals or strings.
This is faster than converting back to a string: 
```
return random.getrandbits(1) + 48
```
so I return ordinals for everything.
Also `str.maketrans()` produces a `dict` with ordinal->ordinal mapping.

File: cocotb/binary.py
Comment (garmin-mjames) [reply cocotb/binary.py]: All instances of each character to be translated will be the same, either `0` or `1`, which is the same as current behavior. I noted this in the commit message.

File: cocotb/binary.py
Comment (ktbarrett) [reply cocotb/binary.py]: #1291

**Comment by garmin-mjames** [comment]
@eric-wieser As implemented, this keeps the current `resolve` behavior (see #1487).
Should I go back to a `re.sub` approach to get true per-bit resolving? It's slightly slower than `str.translate` but much faster than the current approach.

**Comment by eric-wieser** [comment]
Can you put the new example in a standalone PR first?

**Comment by garmin-mjames** [comment]
We should probably resolve #1487 in this PR as well.

**Comment by ktbarrett** [comment]
@garmin-mjames The maintainers decided to deprecate the BinaryValue class and replace it (#2059). We are still deciding on how to go about the whole deprecation process of BinaryValue; and will come to that decision before reviewing/merging this PR so we aren't wasting our or your effort on something that may be immediately dropped. I hope you understand, and please feel free to comment on #2059!

**Comment by ktbarrett** [comment]
@garmin-mjames Are you still interested in getting this in for 1.5? Seems like all the prereq PRs are merged.

**Comment by garmin-mjames** [comment]
Yes. I'll rebase and update this in the coming week.

File: cocotb/handle.py
Comment (themperek) [inline cocotb/handle.py:722]: Any idea if we create BinaryValue once and return a copy (https://docs.python.org/3.5/library/copy.html) ? Not to `__init__`?

Comment (garmin-mjames) [reply cocotb/handle.py:722]: Would that save any time? This already bypasses the slow path of `assign() -> binstr()`, which is most of the time taken up by `BinaryValue.__init__()` when you pass in a value.

It might be slightly faster to change `set_trusted_binstr()` to a classmethod and skip the deprecation checks in `__init__()`.

Comment (themperek) [reply cocotb/handle.py:722]: You are probably be right, it is `assign`.

File: cocotb/binary.py
Comment (ktbarrett) [inline cocotb/binary.py:42]:
```suggestion
class _ResolveTable(dict):
```

File: cocotb/binary.py
Comment (ktbarrett) [inline cocotb/binary.py:82]:
```suggestion
_resolve_table = _ResolveTable()
```

File: cocotb/binary.py
Comment (ktbarrett) [inline cocotb/binary.py:443]:
```suggestion
    def _set_trusted_binstr(self, string):
        """set value of a BinaryValue using a str without doing sanity checking"""
```

File: cocotb/binary.py
Comment (ktbarrett) [inline cocotb/binary.py:445]: If we trust the string do we need this?
```suggestion
```

File: cocotb/handle.py
Comment (ktbarrett) [reply cocotb/handle.py:751]:
```suggestion
        # Skip the permitted characters check as we trust the simulator
        result._set_trusted_binstr(binstr)
```

File: cocotb/binary.py
Comment (garmin-mjames) [reply cocotb/binary.py:445]: good point.

File: cocotb/binary.py
Comment (ktbarrett) [inline cocotb/binary.py:36]: I know this was just moved, but resolving `-` to `0` is wrong. "Don't care" is a synonym for `X`. Something to note in the future.

**Review by ktbarrett — APPROVED** [review]

File: cocotb/binary.py
Comment (garmin-mjames) [reply cocotb/binary.py:36]: #1487 with your comment is still open.

**Comment by garmin-mjames** [comment]
Does this warrant a newsfragment? Behavior is identical, just faster.

**Comment by ktbarrett** [comment]
@garmin-mjames I don't think so. I think it's g2g.

**Comment by garmin-mjames** [comment]
Just updated commit messages.

**Pull request merged**
